### ISSUE TITLE
TypeError: unhashable type: 'Strict'

### ISSUE THREAD

**Issue by DimitriPapadopoulos** [pydantic/pydantic#5377]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

The code below used to work in V1, results in the following error in V2:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/typing.py", line 312, in inner
    return func(*args, **kwds)
  File "/usr/lib/python3.10/typing.py", line 403, in __getitem__
    return self._getitem(self, parameters)
  File "/usr/lib/python3.10/typing.py", line 516, in Union
    parameters = _remove_dups_flatten(parameters)
  File "/usr/lib/python3.10/typing.py", line 281, in _remove_dups_flatten
    return tuple(_deduplicate(params))
  File "/usr/lib/python3.10/typing.py", line 255, in _deduplicate
    all_params = set(params)
  File "/usr/lib/python3.10/typing.py", line 1651, in __hash__
    return hash((self.__origin__, self.__metadata__))
TypeError: unhashable type: 'Strict'
```

### Example Code

```Python
from typing import Literal, Union
from pydantic import StrictStr

VariableList = Union[Literal[1], StrictStr]
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0a1
                 install path: /my/path/pydantic/pydantic
               python version: 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]
                     platform: Linux-5.19.0-38-generic-x86_64-with-glibc2.35
     optional deps. installed: ['typing-extensions']
```

**Comment by DimitriPapadopoulos** [comment]
Definition of `StrictStr`:
https://github.com/pydantic/pydantic/blob/d1b42d5ca2eb62936bd7c84622e9c45ebf494a9a/pydantic/types.py#L194

From there, we can further simplify the example code:

### New Example Code

```python
from typing import Union
from typing_extensions import Annotated
from pydantic import Strict

Union[None, Annotated[str, Strict()]]
```

**Comment by samuelcolvin** [comment]
Thanks so much for reporting this.

should be as simple as adding `def __hash__(self): self.strict` to `Strict`.

@hramezani please can you add this, add a test and make sure other similar types are hashable and therefore work in tests.

I guess we should also check that these work for building a pydantic core schema - e.g. can be used as a field on a model.

**Comment by dmontagu** [comment]
@DimitriPapadopoulos this should be fixed now in `pydantic==2.0a2`. Please let us know if you still have issues.
